File: backend/app/geo.py
"""緯度経度をローカルな平面メートル座標へ変換する小さなヘルパー。

ルート上の各地点を「最初の地点を原点とした相対位置」に置くために使う。
数 km 程度の範囲なら等距円筒近似で十分な精度。

座標規約:
- east  : 東向きを + (m)  → ワールドの +X
- north : 北向きを + (m)  → ワールドの -Z（Three.js/glTF は -Z が前方/北）
"""

import math

# 緯度 1 度あたりのメートル（地球平均）。経度方向は cos(緯度) で縮む。
METERS_PER_DEG_LAT = 110_540.0
METERS_PER_DEG_LNG = 111_320.0


def local_offset_meters(
    origin: tuple[float, float], point: tuple[float, float]
) -> tuple[float, float]:
    """origin を基準とした point の (east_m, north_m) を返す。

    origin, point はいずれも (lat, lng)。
    """
    lat0, lng0 = origin
    lat, lng = point
    north = (lat - lat0) * METERS_PER_DEG_LAT
    east = (lng - lng0) * METERS_PER_DEG_LNG * math.cos(math.radians(lat0))
    return east, north


def world_translation(
    origin: tuple[float, float], point: tuple[float, float]
) -> list[float]:
    """origin を基準にした point のワールド平行移動量 [x, y, z] を返す。

    east → +X、north → -Z（前方が -Z）。高さ(Y)は 0。
    """
    east, north = local_offset_meters(origin, point)
    return [east, 0.0, -north]

File: frontend/lib/geo.ts
// 緯度経度をローカル平面メートルに近似し、道沿いの点列をサンプリングする。
// バックエンドの geo.py と同じ近似（数km程度なら十分）。

export interface LatLng {
  lat: number;
  lng: number;
}

const METERS_PER_DEG_LAT = 110540;
const METERS_PER_DEG_LNG = 111320;

// 始点・終点間の距離（メートル）。
export function distanceMeters(a: LatLng, b: LatLng): number {
  const latMid = ((a.lat + b.lat) / 2) * (Math.PI / 180);
  const north = (b.lat - a.lat) * METERS_PER_DEG_LAT;
  const east = (b.lng - a.lng) * METERS_PER_DEG_LNG * Math.cos(latMid);
  return Math.hypot(north, east);
}

// 始点〜終点を stepMeters 間隔でサンプリングした点列を返す（最大 maxPoints 点）。
export function sampleLine(
  start: LatLng,
  end: LatLng,
  stepMeters: number,
  maxPoints: number,
): LatLng[] {
  const distance = distanceMeters(start, end);
  const count = Math.max(
    2,
    Math.min(maxPoints, Math.round(distance / stepMeters) + 1),
  );
  const points: LatLng[] = [];
  for (let i = 0; i < count; i++) {
    const t = i / (count - 1);
    points.push({
      lat: start.lat + (end.lat - start.lat) * t,
      lng: start.lng + (end.lng - start.lng) * t,
    });
  }
  return points;
}
